### ISSUE TITLE
Set up continuous integration server for GRNsight

### ISSUE THREAD

**Issue by kdahlquist** [dondi/GRNsight#248]
Being inspired by what is shown in the markdown readme on this page:

https://github.com/benbalter/word-to-markdown/blob/master/README.md

We could set up a continuous integration server that listens to the repository and automatically runs tests every time code is pushed to the repository.

This would take us to the "next level".

**Comment by dondi** [comment]
It turns out that GitHub endorses (and integrates with) a good number of “continuous integration providers:”

    https://github.com/integrations/feature/build

When it comes time to actively work on this issue, we can start with that list.

**Comment by kdahlquist** [comment]
The 10 Simple Rules (#356) article recommends [Travis CI](https://travis-ci.org/).

**Comment by eileenchoe** [comment]
Set up Travis CI on my fork to test it out, by adding a .travis.yml config file. Travis CI runs a full build and runs all the tests every time code is pushed to the repository or a pull request is submitted, and has nice UI with build information and history. 

The catch is that since we are working on forks, we will each need to configure Travis CI for each of our forked repositories in order to get builds for each commit. I wanted to reopen the discussion on forks vs. branches (I don't remember last year when I first started why development was switched to forks), because having branches will allow us to use CI tools much more efficiently. Since @dondi is the owner of the GRNsight repo, he will need to open a travis ci account and enable GRNsight.

**Comment by kdahlquist** [comment]
We are OK with going back to branches. @dondi and @eileenchoe will meet between now and the next meeting to accelerate the process of getting Travis CI up and running completely, including linting.

**Comment by kdahlquist** [comment]
Pending the use of CI, we will establish a new branch naming convention where the developer's name is part of their branch in progress. CI will help keep the branches in order; conversely, a developer should periodically merge `beta` into their branch so that things do not diverge for too long.

As a release approaches, we will merge all branches-to-be-released into `beta` and have a validation cycle prior to releasing then merging to `master`.

**Comment by eileenchoe** [comment]
@dondi do you have time on Monday to meet to get the CI running? I am available anytime after 10AM.

**Comment by dondi** [comment]
@eileenchoe On Mondays I have a back-to-back class+meeting+class block that ends at 12:30pm, followed by a 1-2pm Rains LLC office hour, after which I need to go off-campus for carpool stuff.

Thus, if it’s OK with you, we can meet in Rains around 1pm to work on this. After my 12:30pm class, I was planning to pick up some lunch at Einsteins then head to Rains (3rd floor reading room). We can meet at Rains or beforehand at Einsteins if it's useful to talk and sync up a bit before actually sitting to work on CI setup.

**Comment by eileenchoe** [comment]
@dondi yes that works for me! I'll be working at the tables in front of Einsteins in the morning then, and meet up after your 12:30.

**Comment by dondi** [comment]
@eileenchoe Great, I’ll look for you in front of Einsteins when I get there!

**Comment by eileenchoe** [comment]
All set up-- log on to travis-ci.org (using your Github login) to view build histories! Travis builds are automatically started for all commits and pull requests on any GRNsight branch. @ebachoura @yshin4 please pull from beta from your branches so builds from your branches can start being generated by travis

**Comment by eileenchoe** [comment]
All successful travis builds will initiate code coverage with coveralls #374

**Comment by eileenchoe** [comment]
Added badges for build status and coverage for beta to the README.md on the beta branch. Badges are generated by markdown and are updated with each travis-ci build. As @dondi and I discussed, once this code is merged into master, the badges should reflect code coverage and build status for the master branch.

**Comment by kdahlquist** [comment]
This is probably not the right issue for this, but it might be nice to move our "badge" for the doi up to the top of the ReadMe so that all the badges are at the top.

**Comment by ebachoura** [comment]
Alright so I pulled and then logged in to Travis. Everything looks good, except to be honest I don't really understand how to use it and what it's going to be used for.

**Comment by dondi** [comment]
@kdahlquist The DOI badge was easy enough to move, so I went ahead and did it as part of tonight’s round of cleanup and maintenance. It is visible on `beta` and should go on `master` whenever we do a release.

@ebachoura Travis will automatically do builds and tests of all branches of GRNsight and indicate branches that have errors. Your URL may differ, but for me, when I go to https://travis-ci.org/dondi/GRNsight/branches, I can see that most branches are passing except for `yshin4` (one failing test). Thus, even if we forget to run linting and tests on our own, Travis CI will do this for us whenever we push work to GitHub.

This issue is a candidate for closure at our September 13 meeting.

**Comment by kdahlquist** [comment]
done; we have our badges!
